src/ccall.cpp

```diff
@@ -924,91 +759,95 @@ static jl_cgval_t emit_llvmcall(jl_codectx_t &ctx, jl_value_t **args, size_t nar
+        }
+        else {
+            auto Buf = MemoryBuffer::getMemBuffer(
+                StringRef((char *)jl_array_data(ir), jl_array_len(ir)), "llvmcall",
+                /*RequiresNullTerminator*/ false);
+            Expected<std::unique_ptr<Module>> ModuleOrErr =
+                parseBitcodeFile(*Buf, jl_LLVMContext);
+            if (Error Err = ModuleOrErr.takeError()) {
+                std::string Message;
+                handleAllErrors(std::move(Err),
+                                [&](ErrorInfoBase &EIB) { Message = EIB.message(); });
+                std::string message = "Failed to parse LLVM bitcode: \n";
+                raw_string_ostream stream(message);
+                stream << Message;
+                emit_error(ctx, stream.str());
+                return jl_cgval_t();
+            }
+            Mod = std::move(ModuleOrErr.get());
+        }
+
+        Function *f = Mod->getFunction(jl_string_data(entry));
+        if (!f) {
+            emit_error(ctx, "Module IR does not contain specified entry function");
+            return jl_cgval_t();
+        }
+        f->setName(ir_name);
+
+        // verify the function type
         assert(!f->isDeclaration());
         assert(f->getReturnType() == rettype);
         int i = 0;
         for (std::vector<Type *>::iterator it = argtypes.begin();
             it != argtypes.end(); ++it, ++i)
             assert(*it == f->getFunctionType()->getParamType(i));
+    }
 
-        if (f->getParent() != jl_Module) {
-            FunctionMover mover(jl_Module, f->getParent());
-            f = mover.CloneFunction(f);
-        }
+    // copy module properties that should always match
+    Mod->setTargetTriple(jl_Module->getTargetTriple());
+    Mod->setDataLayout(jl_Module->getDataLayout());
 
-        std::string message = "Malformed LLVM function: \n";
-        raw_string_ostream stream(message);
-        if (verifyFunction(*f, &stream)) {
-            emit_error(ctx, stream.str());
-            return jl_cgval_t();
-        }
+    // verify the IR
+    Function *f = Mod->getFunction(ir_name);
+    assert(f);
+    std::string message = "Malformed LLVM function: \n";
+    raw_string_ostream stream(message);
+    if (verifyFunction(*f, &stream)) {
+        emit_error(ctx, stream.str());
+        return jl_cgval_t();
     }
 
-    // Since we dumped all of f's dependencies into the active module,
-    // we cannot reasonably inline it, so leave it there and just emit
-    // a regular call
-    if (!isString) {
-        static int llvmcallnumbering = 0;
-        std::string name;
-        llvm::raw_string_ostream(name) << "jl_llvmcall" << llvmcallnumbering++;
-        f->setName(name);
-        f = prepare_llvmcall(jl_Module, llvmcall_proto(f));
-    }
-    else {
-        f->setLinkage(GlobalValue::LinkOnceODRLinkage);
+    if (Linker::linkModules(*jl_Module, std::move(Mod))) {
+        emit_error(ctx, "Failed to link LLVM bitcode");
+        return jl_cgval_t();
```

Comment:
The state of the runtime is now corrupt here. This needs to call `abort` instead.